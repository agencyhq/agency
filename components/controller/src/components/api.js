const http = require('http')
const util = require('util')

const metrics = require('@agencyhq/agency-metrics')
const cors = require('cors')
const express = require('express')
const log = require('loglevel')
const morgan = require('morgan')
const passport = require('passport')
const bearer = require('passport-http-bearer')

const pubsub = require('../pubsub')
const router = require('../rest/router')
const RPCServer = require('../rpc/server')

log.setLevel(process.env.LOG_LEVEL || 'info')

const {
  PORT = 3000,
  METRICS = false
} = process.env

const USERS = {
  alltoken: ['enykeev', { scopes: ['all'] }],
  sensortoken: ['enykeev', { scopes: ['sensor'] }],
  ruletoken: ['enykeev', { scopes: ['rule'] }],
  executiontoken: ['enykeev', { scopes: ['execution'] }],
  sometoken: ['someguy', { scopes: ['web'] }]
}

async function handleExecution (rpc, msg) {
  const execution = JSON.parse(msg.content.toString())
  metrics.countExecutions(execution)
  log.debug('reciving %s: %s', msg.fields.routingKey, util.inspect(execution))

  rpc.notify('execution', execution)

  log.debug('acknowledge reciving %s: %s', msg.fields.routingKey, execution.id)
  pubsub.channel.ack(msg)
}

async function handleTrigger (rpc, msg) {
  const trigger = JSON.parse(msg.content.toString())
  metrics.countTriggers(trigger)
  log.debug('reciving %s: %s', msg.fields.routingKey, util.inspect(trigger))

  if (rpc.hasSubscribers('trigger')) {
    rpc.notify('trigger', trigger, { random: true })

    log.debug('acknowledge reciving %s: %s', msg.fields.routingKey, trigger.id)
    pubsub.channel.ack(msg)
  } else {
    log.debug('reject reciving %s: %s', msg.fields.routingKey, trigger.id)
    pubsub.channel.nack(msg)
  }
}

async function handleRule (rpc, msg) {
  const rule = JSON.parse(msg.content.toString())
  // metrics.countTriggers(trigger)
  log.debug('reciving %s: %s', msg.fields.routingKey, util.inspect(rule))

  rpc.notify('rule', rule)

  log.debug('acknowledge reciving %s: %s', msg.fields.routingKey, rule.id)
  pubsub.channel.ack(msg)
}

async function auth (token) {
  if (USERS[token]) {
    const [username, info] = USERS[token]
    return {
      username,
      info
    }
  }
}

passport.use(new bearer.Strategy(
  (token, done) => {
    auth(token)
      .then(({ username, info } = {}) => done(null, username || false, info))
      .catch(err => done(err))
  }
))

async function main () {
  await pubsub.init()

  const authenticate = async ({ token }) => {
    const identity = await auth(token)

    if (!identity) {
      return false
    }

    const { username: user, info: { scopes } } = identity

    return {
      user,
      scopes
    }
  }

  const app = express()
  const server = http.createServer(app)
  const rpc = new RPCServer({ server, authenticate })

  if (METRICS) {
    app.use(metrics.middleware({
      prefix: 'ifttt_api_'
    }))
  }
  app.use(cors())
  app.use(morgan('combined'))
  app.use(passport.authenticate('bearer', { session: false }))
  app.use(express.json())
  app.use(router('../openapi.yaml'))

  pubsub.subscribe('execution', msg => handleExecution(rpc, msg), {
    name: false,
    exclusive: true
  })

  pubsub.subscribe('rule', msg => handleRule(rpc, msg), {
    name: false,
    exclusive: true
  })

  rpc.registerMethod('ready', () => {
    if (pubsub.isSubscribed('trigger')) {
      return
    }

    pubsub.subscribe('trigger', msg => handleTrigger(rpc, msg))
  }, {
    scopes: ['any']
  })

  rpc.on('disconnect', () => {
    if (!rpc.hasSubscribers('trigger')) {
      pubsub.unsubscribe('trigger')
    }
  })

  server.listen(PORT, () => {
    log.info(`Listening on http://localhost:${PORT}`)
  })
}

main()
  .catch(e => {
    log.error(e)
    process.exit(1)
  })
