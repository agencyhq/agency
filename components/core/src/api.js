const http = require('http')
const util = require('util')

const metrics = require('@agencyhq/agency-metrics')
const cors = require('cors')
const express = require('express')
const log = require('loglevel')
const morgan = require('morgan')
const passport = require('passport')
const bearer = require('passport-http-bearer')

const pubsub = require('./pubsub')
const router = require('./rest/router')
const RPCServer = require('./rpc/server')
const models = require('./models')

log.setLevel(process.env.LOG_LEVEL || 'info')

const {
  PORT = 3000,
  METRICS = false
} = process.env

async function handleMessage (type, rpc, msg) {
  const execution = JSON.parse(msg.content.toString())
  metrics.countMessages(type, execution)
  log.debug('reciving %s: %s', msg.fields.routingKey, util.inspect(execution))

  await rpc.notify(type, execution)

  log.debug('acknowledge reciving %s: %s', msg.fields.routingKey, execution.id)
  pubsub.channel.ack(msg)
}

async function auth (token) {
  const identity = await models.Tokens
    .forge({ id: token || '' })
    .fetch({ require: false })
    .then(model => model ? model.toJSON() : {})
    .catch(err => {
      log.warn(err)
      return {}
    })

  const {
    id,
    user,
    meta: {
      scopes
    } = {}
  } = identity

  if (id === token) {
    return {
      user,
      scopes
    }
  }

  return false
}

passport.use(new bearer.Strategy(
  (token, done) => {
    auth(token)
      .then(({ user, scopes } = {}) => done(null, user || false, { scopes }))
      .catch(err => done(err))
  }
))

async function main () {
  await pubsub.init()

  const authenticate = async ({ token }) => {
    const identity = await auth(token)

    if (!identity) {
      return false
    }

    const {
      user,
      scopes
    } = identity

    return {
      user,
      scopes
    }
  }

  const app = express()
  const server = http.createServer(app)
  const rpc = new RPCServer({ server, authenticate })

  if (METRICS) {
    app.use(metrics.middleware({
      prefix: 'ifttt_api_'
    }))
  }
  app.use(cors())
  app.use(morgan('short', {
    stream: {
      write: str => log.debug(str)
    }
  }))
  app.use(passport.authenticate('bearer', { session: false }))
  app.use(express.json())
  app.use(router('../openapi.yaml', opId => {
    if (!opId) {
      return () => {
        throw new Error('not implemented')
      }
    }

    return require(`./rest/methods/${opId}`)
  }))

  pubsub.subscribe('execution', msg => handleMessage('execution', rpc, msg), {
    name: false,
    exclusive: true
  })

  pubsub.subscribe('rule', msg => handleMessage('rule', rpc, msg), {
    name: false,
    exclusive: true
  })

  pubsub.subscribe('trigger', msg => handleMessage('trigger', rpc, msg), {
    name: false,
    exclusive: true
  })

  pubsub.subscribe('result', msg => handleMessage('result', rpc, msg), {
    name: false,
    exclusive: true
  })

  rpc.registerMethod('ready', () => {}, { scopes: ['any'] })

  rpc.on('disconnect', () => {
    if (!rpc.hasSubscribers('trigger')) {
      pubsub.unsubscribe('trigger')
    }
  })

  server.listen(PORT, () => {
    log.info(`Listening on http://localhost:${PORT}`)
  })
}

main()
  .catch(e => {
    log.error(e)
    process.exit(1)
  })
