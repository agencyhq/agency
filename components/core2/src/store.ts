import { Operation } from './models/operation';

export class Storage {
  private operations: Operation[] = [];

  loadOperations(operations: Operation[]): void {
    this.operations = [...operations];
  }

  getOperations(): Operation[] {
    return this.operations;
  }

  getOperation(id: string): Operation | undefined {
    return this.operations.find((operation) => operation.id === id);
  }

  createOperation(operation: Operation): Operation {
    this.operations.push(operation);
    return operation;
  }

  updateOperation(operation: Operation): Operation {
    const index = this.operations.findIndex((op) => op.id === operation.id);
    this.operations[index] = operation;
    return operation;
  }

  deleteOperation(id: string): Operation | undefined {
    const index = this.operations.findIndex((op) => op.id === id);
    if (index === -1) {
      return undefined;
    }
    return this.operations.splice(index, 1)[0];
  }

  clearOperations(): void {
    this.operations = [];
  }
}
