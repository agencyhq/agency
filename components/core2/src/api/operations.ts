import { Router } from 'express';
import { v4 as uuidv4 } from 'uuid';

import { Operation } from '../models/operation';
import { Storage } from '../store';
import { validateStatus } from '../utils';

export function operationsApi(app: Router, storage: Storage) {
  /**
   * Operations
   *
   * GET /operations
   * GET /operations/:id
   * POST /operations
   * PUT /operations/:id
   * DELETE /operations/:id
   * GET /operations/:id/result
   *
   * @example
   * $ curl -X GET http://localhost:3000/operations
   */
  app.get('/operations', (req, res) => {
    const operations: Operation[] = storage.getOperations();

    res.json(operations);
  });

  /**
   * @example
   * $ curl -X GET http://localhost:3000/operations/1
   */
  app.get('/operations/:id', (req, res) => {
    const id = req.params.id;
    const operation: Operation = storage.getOperation(id);

    if (!operation) {
      res.status(404).send('Operation not found');
      return;
    }

    res.json(operation);
  });

  /**
   * @example
   * $ curl -X POST http://localhost:3000/operations \
   *  -H "Content-Type: application/json" \
   *  -d '{"action":"create","params":{"name":"John Doe"}}'
   */
  app.post('/operations', (req, res) => {
    const operation: Operation = req.body;
    const id = uuidv4();

    operation.id = id;
    operation.status = 'requested';
    operation.createdAt = new Date();
    operation.updatedAt = new Date();

    storage.createOperation(operation);

    res.json(operation);
  });

  /**
   * @example
   * $ curl -X PUT http://localhost:3000/operations/1 \
   *   -H "Content-Type: application/json" \
   *   -d '{"action":"update","params":{"id":"1","name":"Jane Doe"}}'
   */
  app.put('/operations/:id', (req, res) => {
    const id = req.params.id;
    const newOperation: Operation = req.body;

    const operation: Operation = storage.getOperation(id);

    const result = storage.updateOperation({
      ...operation,
      ...newOperation,
      id,
      updatedAt: new Date(),
    });

    res.json(result);
  });

  /**
   * @example
   * $ curl -X DELETE http://localhost:3000/operations/1
   */
  app.delete('/operations/:id', (req, res) => {
    const id = req.params.id;

    storage.deleteOperation(id);

    res.json({ id: req.params.id });
  });

  /**
   * @example
   * $ curl -X GET http://localhost:3000/operations/1/result
   */
  app.get('/operations/:id/result', (req, res) => {
    const id = req.params.id;
    const operation: Operation = storage.getOperation(id);

    if (!operation) {
      res.status(404).send('Operation not found');
      return;
    }

    res.json(operation.result);
  });

  /**
   * @example
   * $ curl -X POST http://localhost:3000/operations/1/claim
   */
  app.post('/operations/:id/claim', (req, res) => {
    const id = req.params.id;
    const operation: Operation = storage.getOperation(id);

    if (!operation) {
      res.status(404).send('Operation not found');
      return;
    }

    if (operation.status !== 'requested') {
      res.status(400).send('Operation is not in requested state');
      return;
    }

    const result = storage.updateOperation({
      ...operation,
      status: 'claimed',
      updatedAt: new Date(),
    });

    res.json(result);
  });

  /**
   * @example
   * $ curl -X GET http://localhost:3000/operations/1/status
   */
  app.get('/operations/:id/status', (req, res) => {
    const id = req.params.id;
    const operation: Operation = storage.getOperation(id);

    if (!operation) {
      res.status(404).send('Operation not found');
      return;
    }

    res.json({ status: operation.status });
  });

  /**
   * @example
   * $ curl -X POST http://localhost:3000/operations/1/status \
   *  -H "Content-Type: application/json" \
   *  -d '{"status":"processing","result":{"name":"John Doe"}}'
   */
  app.post('/operations/:id/status', (req, res) => {
    const id = req.params.id;
    const operation: Operation = storage.getOperation(id);

    if (!operation) {
      res.status(404).send('Operation not found');
      return;
    }

    if (!['claimed', 'processing'].includes(operation.status)) {
      res.status(400).send('Operation is not in claimed or processing state');
      return;
    }

    const { status, result } = req.body;

    if (!['processing', 'completed', 'failed'].includes(validateStatus(status))) {
      res.status(400).send('Operation can only transition to processing, completed or failed state');
      return;
    }

    const op = storage.updateOperation({
      ...operation,
      status,
      result: result,
      updatedAt: new Date(),
    });

    res.json(op);
  });
}
