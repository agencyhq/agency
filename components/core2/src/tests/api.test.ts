import { beforeEach, describe, expect, it, jest } from '@jest/globals';
import express from 'express';
import request from 'supertest';

import { operationsApi } from '../api';
import { Storage } from '../store';
import fixtures from './__fixtures__/operations';

describe('api', () => {
  describe('operationsApi', () => {
    let app: express.Express;
    let storage: Storage;

    beforeEach(() => {
      jest.resetAllMocks();
      storage = new Storage();

      storage.loadOperations(fixtures);

      app = express();
      app.use(express.json());
    });

    it('should return operations', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations')
        .expect(200)
        .then((response) => {
          response.body.forEach((operation: any) => {
            expect(operation).toMatchSnapshot({
              createdAt: expect.any(String),
              updatedAt: expect.any(String),
            });
          });
        });
    });

    it('should return operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/1')
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          });
        });
    });

    it('should create operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations')
        .send({ action: 'create', params: { name: 'John Doe' } })
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot({
            id: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          });
        })
        .catch((err) => {
          console.log(err);
          throw err;
        });
    });

    it('should update operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .put(`/operations/1`)
        .send({ action: 'update', params: { name: 'Jane Doe' } })
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          });
        });
    });

    it('should delete operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .delete('/operations/1')
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return 404 on non-existent operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/5')
        .expect(404)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should claim operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/3/claim')
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          });
        });
    });

    it('should return 404 on non-existent operation claim', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/5/claim')
        .expect(404)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('shoult return 400 on already claimed operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/2/claim')
        .expect(400)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return result of operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/1/result')
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return 404 on non-existent operation result', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/5/result')
        .expect(404)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return status of operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/1/status')
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return 404 on non-existent operation status', async () => {
      operationsApi(app, storage);

      await request(app)
        .get('/operations/5/status')
        .expect(404)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should post status of operation', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/2/status')
        .send({ status: 'processing' })
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          });
        });
    });

    it('should return 404 on non-existent operation status post', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/5/status')
        .send({ status: 'processing' })
        .expect(404)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return 400 on invalid status operation is in', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/1/status')
        .send({ status: 'failed' })
        .expect(400)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should return 400 on invalid status operation should transition to', async () => {
      operationsApi(app, storage);

      await request(app)
        .post('/operations/2/status')
        .send({ status: 'requested' })
        .expect(400)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });
    
  });
});
