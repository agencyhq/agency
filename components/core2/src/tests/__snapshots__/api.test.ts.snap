// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api operationsApi should claim operation 1`] = `
{
  "action": "delete",
  "createdAt": Any<String>,
  "id": "3",
  "params": {
    "id": "1",
  },
  "result": {
    "id": "1",
  },
  "status": "claimed",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should create operation 1`] = `
{
  "action": "create",
  "createdAt": Any<String>,
  "id": Any<String>,
  "params": {
    "name": "John Doe",
  },
  "status": "requested",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should delete operation 1`] = `
{
  "id": "1",
}
`;

exports[`api operationsApi should post status of operation 1`] = `
{
  "action": "update",
  "createdAt": Any<String>,
  "id": "2",
  "params": {
    "id": "1",
    "name": "Jane Doe",
  },
  "status": "processing",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should return 400 on invalid status operation is in 1`] = `{}`;

exports[`api operationsApi should return 400 on invalid status operation should transition to 1`] = `{}`;

exports[`api operationsApi should return 404 on non-existent operation 1`] = `{}`;

exports[`api operationsApi should return 404 on non-existent operation claim 1`] = `{}`;

exports[`api operationsApi should return 404 on non-existent operation result 1`] = `{}`;

exports[`api operationsApi should return 404 on non-existent operation status 1`] = `{}`;

exports[`api operationsApi should return 404 on non-existent operation status post 1`] = `{}`;

exports[`api operationsApi should return operation 1`] = `
{
  "action": "create",
  "createdAt": Any<String>,
  "id": "1",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "1",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should return operations 1`] = `
{
  "action": "create",
  "createdAt": Any<String>,
  "id": "1",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "1",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should return operations 2`] = `
{
  "action": "update",
  "createdAt": Any<String>,
  "id": "2",
  "params": {
    "id": "1",
    "name": "Jane Doe",
  },
  "result": {},
  "status": "claimed",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should return operations 3`] = `
{
  "action": "delete",
  "createdAt": Any<String>,
  "id": "3",
  "params": {
    "id": "1",
  },
  "result": {
    "id": "1",
  },
  "status": "requested",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi should return result of operation 1`] = `
{
  "id": "1",
  "name": "John Doe",
}
`;

exports[`api operationsApi should return status of operation 1`] = `
{
  "status": "completed",
}
`;

exports[`api operationsApi should update operation 1`] = `
{
  "action": "update",
  "createdAt": Any<String>,
  "id": "1",
  "params": {
    "name": "Jane Doe",
  },
  "result": {
    "id": "1",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`;

exports[`api operationsApi shoult return 400 on already claimed operation 1`] = `{}`;
